application/bsdd_utils.py

```diff
@@ -42,4 +50,64 @@ def get_inst(instance_id):
     
     return hierarchical_bsdd_results     
 
+def get_processed_bsdd_table(bsdd_results, session, schema):
+    """
+    The function returns data for the 'bsdd data' table in a file metrics report. 
+    More specifically,it returns a list of dictionaries containing the classifications (ifc instances, in this case) 
+    and their valid/invalid counts.
+    """
+    bsdd_instances = [bsdd_table(result, session, schema) for result in bsdd_results]
+    bsdd_data = defaultdict(lambda: {'valid': 0, 'invalid': 0, 'domain_source': 'classification not found'})
+
+    for item in bsdd_instances:
+        classification = item['classification']
+        validity = item['validity']
+        bsdd_data[classification][validity] += 1
+
+    bsdd_data = [{**{'classification': k}, **v} for k, v in bsdd_data.items()]
+    return bsdd_data
+
+def instance_supertypes(observed_type, schema):
+    allowed_types = [observed_type]
+
+    while True:
+        try:
+            result = (lambda x: ifcopenshell.ifcopenshell_wrapper.schema_by_name(schema).declaration_by_name(x).supertype().name())(allowed_types[-1])
+        except AttributeError: #NoneType
+            break
+
+        allowed_types.append(result)
+    return allowed_types
+
+def bsdd_table(bsdd_result, session, schema):
+    inst = get_inst(session, bsdd_result['instance_id'])
+    observed_type = inst.ifc_type
+    required_type = bsdd_result['bsdd_type_constraint']
+    domain_source = bsdd_result['bsdd_classification_uri']
+    validity = "valid" if required_type in instance_supertypes(observed_type, schema) else 'invalid'
```

Comment:
1) We don't determine the validity of a bsdd_result only with respect to the type
2) To check if a bsdd_result object is valid, the following attributes of the object need to be set to 1: val_ifc_type, val_property_set, val_property_name, val_property_type, val_property_value

Reply:
https://github.com/Ghesselink/ifc-pipeline-validation/commit/19deaf9e934d8e6e79445b9bf86657cf68aeba32